Make error reporting match that of SBT's

jadenPete opened the pull request:
This PR fixes the following discrepancy with our compiler error reporting and SBT's:

Ours:
```
[Error] src/main/scala/higherkindness/rules_scala/workers/zinc/test/TestRunner.scala:233 type mismatch;
[Error]  found   : Seq[higherkindness.rules_scala.common.sbt_testing.TestDefinition]
[Error]  required: List[higherkindness.rules_scala.common.sbt_testing.TestDefinition]
one error found
```

Using SBT or `scalac`:
```
src/main/scala/higherkindness/rules_scala/workers/zinc/test/TestRunner.scala:233: error: type mismatch;
 found   : Seq[higherkindness.rules_scala.common.sbt_testing.TestDefinition]
 required: List[higherkindness.rules_scala.common.sbt_testing.TestDefinition]
          runner.execute(filteredTests, testScopeAndName.getOrElse(""), testRunnerArgs.frameworkArgs)
                         ^
1 error
```

The `[Error] ` line prefix is added by us, but the detailed position information is missing.

Additionally, this PR fixes the test failures on `lucid-master`.

Commit:
Don't use a custom formatter for compilation errors

Commit:
Made ZincRunner relativize SemanticDB paths to the sandbox directory

jjudd commented on src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala:
```diff
@@ -49,6 +49,19 @@ object Analysis {
+    def adjustStringPath(path: String) = SandboxUtil.getSandboxPath(workDir, Paths.get(path)).toString
+
+    options.flatMap {
+      case s"-P:semanticdb:targetroot:$path" =>
```
Are there other other common compiler options that we know of which need a specific path?

This is a tricky issue because in theory there are infinite options which could do so. Even better yet they get specified by the user.

I suppose it's probably just not work thinking about for now.

jadenPete replied:
> This is a tricky issue because in theory there are infinite options which could do so. Even better yet they get specified by the user.

I thought of this when I made this change and did feel slimy modifying compiler options like this, but I think it's better to modify compiler options than pass SemanticDB options separately. My reasoning is that because there are likely many compiler options that reference paths and some of those compiler options may be used outside of this ruleset (by those defining Scala libraries, binaries, tests, etc..), it's much less confusing to intercept compiler options than direct users to use another set of options instead.

I could go through every compiler option that references file paths, but that list is probably endless and I figured we can expand this list as necessary.

jjudd commented on src/main/scala/higherkindness/rules_scala/workers/common/LoggedReporter.scala:
```diff
@@ -2,18 +2,18 @@ package higherkindness.rules_scala
-    } else {
-      problem.rendered().orElse(problem.toString)
-    }
+    if (shouldEnhanceMessage) colorFunction(formattedProblem) else formattedProblem
```
Does this change how the Scala 3 messages look? Because those were very nice before.

jadenPete replied:
I don't think it should. Before, we were calling `Problem#rendered` and defaulting to `Problem#toString`. `ProblemStringFormats` also attempts to call `Problem#rendered`, but defaults a message that's much more detailed than `Problem#toString`
https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavaErrorParser.scala#L59
https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ProblemStringFormats.scala#L37

I think the niceness you're referring to is the product of `Problem#rendered`.

jjudd replied:
I checked out this branch and a branch without this change. Confirmed they're still the same.

jjudd commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:
```diff
@@ -52,14 +52,12 @@ object ZincRunnerWorkerConfig {
  * Zinc has two caches:
  *   1. a ClassLoaderCache which is a soft reference cache for classloaders of Scala compilers.
- *   2. a CompilerCache which is a hard reference cache for (I think) Scala compiler instances.
+ *   1. a CompilerCache which is a hard reference cache for (I think) Scala compiler instances.
```
I think this should be 2 instead of 1?

jadenPete replied:
It's supposed to be `1`. The reason is was formatting incorrectly before is because, confusingly, every item in a Scaladoc list should begin with `1. `:
https://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html#markup

jjudd commented on tests/zinc/ErrorReporting.scala:
```diff
@@ -0,0 +1,5 @@
+object Main {
+  private def printZooAnimals(animals: List[String]): Unit = println(s"The zoo has ${animals.mkString(",")}")
```
Can we add a Scala 3 test for error reporting to make sure the nice Scala 3 error reporting works well still?

jadenPete replied:
Sure thing!

Commit:
Fixed the compile/logger test

Not only does ProblemStringFormats format problems (compilation
messages) slightly differently, but it uses `xsbti.Position#pointer`
rather than `xsbti.Position#offset` as the column number.

Prior to v1.10.1 of Zinc, `sbt.internal.inc.javac.JavaPosition`, a
concrete implementation of `Position`, set `pointer` to `None`
unconditionally, meaning that `ProblemStringFormats` didn't have access
to the line number. To fix this, I upgraded our version of Zinc to
v1.10.1 from v1.10.0.
